scripts/suite.py

```diff
@@ -29,6 +30,16 @@
   <config_yaml>               Optional extra job yaml to include
   -s <suite>, --suite <suite>
                               The suite to schedule
+  -C <suite_path>, --calc <suite_path>
+                              Path to a suite directory for calculation,
+                              see also --desc, --list, --matrix, as well as 
+                              options --subset, --filter and --filter-out
+  -P, --desc                  Print job descriptions for the suite,
```

Comment:
I used the cli to understand better, -P and -L are similar to --dry-run, are you saying dry-run doesnt work with local dirs and hence this change?  Is it worth fixing dry-run to handle this, my concern is two cli's doing almost similar job.

Reply:
@vasukulkarni From my understanding --dry-run is to go throw the run, like run but do not schedule a job, correspondingly it generates a lot of noise in the logs, the idea of calc is to drop the noise and focus on suite analysis only.

Reply:
I probably think you are little bit misunderstanding the -dry-run option using same seed

Here is what I tried(dry-run also takes --suite-dir option so it can work with local dirs)

a) ./teuthology-suite --dry-run -s fs -c master  --suite-dir ~/ceph/ceph/  -m vps --seed 3550
b)./teuthology-suite --calc ~/ceph/ceph/qa/suites/fs -P --seed 3550

-L would just print the number of tests and in both cases is easy to figure out
-P has less noise but the yaml snippets are exactly the same, but when the number grows bit large its upto user to make sense, but should be possible to fix that noise better?

-M is a new option in this change.

Reply:
@vasukulkarni I was able to run --dry-run without extra teuthology configured file only on working freshly deployed system. And what I get is not what I expect to have:
1) For me it is not good thing that just for investigation of a suite I need to add noisy arguments like machine type, branch repos, as a user I want to have simple way, and not afraid that if I miss for example one of --ceph, --suite, or --machine (have no idea why I need to enter all those parameters to examine a suite). Probably for someone else it useable.

2) Noisy output, where each facet printed in one line as a debug log, it is not a convenient way of representing data and I don't like there idea to keep in memory the best way to parse to understand whole this:
```
2019-01-23 01:05:25,672.672 INFO:teuthology.suite.util:/home/ubuntu/teuthology/virtualenv/bin/teuthology-schedule --name ubuntu-2019-01-23_01:05:21-rbd-ses6---basic-openstack --num 1 --worker openstack --dry-run --priority 1000 --description 'rbd/cli_v1/{base/install.yaml clusters/{fixed-1.yaml openstack.yaml} features/format-1.yaml msgr-failures/few.yaml objectstore/bluestore-bitmap.yaml pool/small-cache-pool.yaml supported-random-distro$/rhel_latest.yaml workloads/rbd_cli_import_export.yaml}' -- /tmp/schedule_suite_cUtuwe src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/base/install.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/clusters/fixed-1.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/clusters/openstack.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/features/format-1.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/msgr-failures/few.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/objectstore/bluestore-bitmap.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/pool/small-cache-pool.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/supported-random-distro$/rhel_latest.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/workloads/rbd_cli_import_export.yaml
2019-01-23 01:05:25,672.672 INFO:teuthology.suite.run:Scheduling rbd/mirror/{base/install.yaml cluster/{2-node.yaml openstack.yaml} msgr-failures/few.yaml objectstore/bluestore-comp.yaml supported-random-distro$/centos_latest.yaml users/mirror.yaml workloads/rbd-mirror-workunit-config-key.yaml}
```
Like I need to add always '2>&1 | grep -v teuthology-schedule ... | sed ...' or 'grep Scheduling' it complicates everything so much.
3) It does extra steps, which I don't like to be executed anyway while working on local copy, like, each try to run teuthology-suite:

```
2019-01-23 01:22:10,102.102 INFO:teuthology.suite.run:ceph sha1: 237cb653cce794c517680eb53da8045214f4e7b1
2019-01-23 01:22:10,103.103 INFO:teuthology.suite.run:skipping ceph package verification
2019-01-23 01:22:10,103.103 INFO:teuthology.suite.run:teuthology branch: master
2019-01-23 01:22:11,492.492 INFO:teuthology.suite.run:ceph branch: ses6 237cb653cce794c517680eb53da8045214f4e7b1
```
and it is in the best case, when you have correctly setup the client node with all configs.

Reply:
> @vasukulkarni I was able to run --dry-run without extra teuthology configured file only on working freshly deployed system. And what I get is not what I expect to have:
> 
> 1. For me it is not good thing that just for investigation of a suite I need to add noisy arguments like machine type, branch repos, as a user I want to have simple way, and not afraid that if I miss for example one of --ceph, --suite, or --machine (have no idea why I need to enter all those parameters to examine a suite). Probably for someone else it useable.
What I am saying is since the output yamls are exactly same with few additional info(which I really dont consider as noise, since the user is expected few details when working with teuthology) it might be better we standardize on same options rather than new sub options, probably add --quiet option, I am not saying we use crazy cli options to filter out things we dont want, it should be easy to fix this in the suite function and silence the logs we dont need. It *will* be helpful for everybody if we fix the standard options everyone is already using and simplify things if they are doing same things.

> 2. Noisy output, where each facet printed in one line as a debug log, it is not a convenient way of representing data and I don't like there idea to keep in memory the best way to parse to understand whole this:
> 
> ```
> 2019-01-23 01:05:25,672.672 INFO:teuthology.suite.util:/home/ubuntu/teuthology/virtualenv/bin/teuthology-schedule --name ubuntu-2019-01-23_01:05:21-rbd-ses6---basic-openstack --num 1 --worker openstack --dry-run --priority 1000 --description 'rbd/cli_v1/{base/install.yaml clusters/{fixed-1.yaml openstack.yaml} features/format-1.yaml msgr-failures/few.yaml objectstore/bluestore-bitmap.yaml pool/small-cache-pool.yaml supported-random-distro$/rhel_latest.yaml workloads/rbd_cli_import_export.yaml}' -- /tmp/schedule_suite_cUtuwe src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/base/install.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/clusters/fixed-1.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/clusters/openstack.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/features/format-1.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/msgr-failures/few.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/objectstore/bluestore-bitmap.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/pool/small-cache-pool.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/supported-random-distro$/rhel_latest.yaml src/github.com_suse_ceph_ses6/qa/suites/rbd/cli_v1/workloads/rbd_cli_import_export.yaml
> 2019-01-23 01:05:25,672.672 INFO:teuthology.suite.run:Scheduling rbd/mirror/{base/install.yaml cluster/{2-node.yaml openstack.yaml} msgr-failures/few.yaml objectstore/bluestore-comp.yaml supported-random-distro$/centos_latest.yaml users/mirror.yaml workloads/rbd-mirror-workunit-config-key.yaml}
> ```
> Like I need to add always '2>&1 | grep -v teuthology-schedule ... | sed ...' or 'grep Scheduling' it complicates everything so much.
> 3) It does extra steps, which I don't like to be executed anyway while working on local copy, like, each try to run teuthology-suite:
> 
> ```
> 2019-01-23 01:22:10,102.102 INFO:teuthology.suite.run:ceph sha1: 237cb653cce794c517680eb53da8045214f4e7b1
> 2019-01-23 01:22:10,103.103 INFO:teuthology.suite.run:skipping ceph package verification
> 2019-01-23 01:22:10,103.103 INFO:teuthology.suite.run:teuthology branch: master
> 2019-01-23 01:22:11,492.492 INFO:teuthology.suite.run:ceph branch: ses6 237cb653cce794c517680eb53da8045214f4e7b1
> ```
> and it is in the best case, when you have correctly setup the client node with all configs.

Reply:
@vasukulkarni 
I see, and want to admit that, it would be great to have some kind of standardization, the only think I don't like is to use `--dry-run` command for just listing a suite, just because the purpose of dry-run is different, as a confirmation of it, --dry-run does not prints out the suites in pretty format it dumps the command logs (info or debug), like dump `teuthology-schedule` command to stderr. And it is just coincidence that all those yaml files have got into debug dump. In other words using dry-run for calc purpose looks like workaround. I would agree that instead `--calc` we would rather use `--suite-dir` however I would prefer it to point to the directory of a particular suite, instead of pointing to the root of some repo with hardcoded file structure. Summarizing using debug like output with all those timestamps and noisy decorations  just for investigating suites looks very weird and clumsy.

Reply:
I didnt' get what you meant by pointing at root dir, I can point at suite level and it works for me. 

If we want to keep this outside `--dry-run`, then how about using a better name instead of --calc (like --analyze) ?

and instead of calculation use analysis in description below,  Also -a option is available.

```
  -C <suite_path>, --calc <suite_path>
                              Path to a suite directory for calculation,
                              see also --desc, --list, --matrix, as well as 
                              options --subset, --filter and --filter-out
```

Reply:
Also -P and -L should be default option.

Reply:
@vasukulkarni The way I use this command is to evaluate the capacity of a suite, so, the default behaviour just to look at the number of cases for the suite, so I don't need grep out all the cases. And after playing with filters, the is a sense to print out the description (-P option). Btw, this is why I used 'calc' keyword, short for calculate, however I don't mind to rename it to something else, but I would to leave a possibility to enter it multiple times, since it is useful sometimes. I also prefer short command name over the longer, like 'teuthology-describe-tests', As for -L options it will also be overkill because the list of yaml is the same for many facets. I don't like it be enable by default, and by the way to disable will require more complicated command line combination either: '-L false' or 'grep -V ...'

Reply:
> I didnt' get what you meant by pointing at root dir, I can point at suite level and it works for me.
I mean 'root' git repository dir, see example:
1) You have to provide '--ceph', '--suite', '--machine' (why should I?):
```
$ teuthology-suite --dry-run --suite-dir ~/src/github.com_SUSE_ceph_ses5/qa/suites/rbd/basic
usage: teuthology-suite --help
       teuthology-suite [-v | -vv ] --ceph <ceph> --suite <suite> [options] [<config_yaml>...]
       teuthology-suite [-v | -vv ] --ceph <ceph> --rerun <name>  [options] [<config_yaml>...]
```
2) You have to provide suite name explicitly as separate argument, and even after that it looks for specific sub-directory of '--suite-dir', in particular 'qa/suites':
```

$ teuthology-suite --dry-run --suite-dir ~/src/github.com_SUSE_ceph_ses5/qa/suites/rbd/basic --ceph ses5 --machine openstack --suite rbd/basic
2019-01-29 14:30:22,731.731 INFO:teuthology.suite.run:ceph sha1: b10c702661a31c8563b3421d6d664de93a0cb0e2
2019-01-29 14:30:22,732.732 INFO:teuthology.suite.run:skipping ceph package verification
2019-01-29 14:30:22,733.733 INFO:teuthology.suite.run:teuthology branch: master
2019-01-29 14:30:24,452.452 INFO:teuthology.suite.run:ceph branch: ses5 b10c702661a31c8563b3421d6d664de93a0cb0e2
Traceback (most recent call last):
  File "/home/ubuntu/teuthology/virtualenv/bin/teuthology-suite", line 11, in <module>
    load_entry_point('teuthology', 'console_scripts', 'teuthology-suite')()
  File "/home/ubuntu/teuthology/scripts/suite.py", line 142, in main
    return teuthology.suite.main(args)
  File "/home/ubuntu/teuthology/teuthology/suite/__init__.py", line 88, in main
    run.prepare_and_schedule()
  File "/home/ubuntu/teuthology/teuthology/suite/run.py", line 311, in prepare_and_schedule
    num_jobs = self.schedule_suite()
  File "/home/ubuntu/teuthology/teuthology/suite/run.py", line 478, in schedule_suite
    build_matrix(suite_path, subset=self.args.subset)
  File "/home/ubuntu/teuthology/teuthology/suite/build_matrix.py", line 50, in build_matrix
    mat, first, matlimit = _get_matrix(path, subset)
  File "/home/ubuntu/teuthology/teuthology/suite/build_matrix.py", line 68, in _get_matrix
    mat = _build_matrix(path)
  File "/home/ubuntu/teuthology/teuthology/suite/build_matrix.py", line 77, in _build_matrix
    raise IOError('%s does not exist (abs %s)' % (path, os.path.abspath(path)))
IOError: /home/ubuntu/src/github.com_SUSE_ceph_ses5/qa/suites/rbd/basic/qa/suites/rbd/basic does not exist (abs /home/ubuntu/src/github.com_SUSE_ceph_ses5/qa/suites/rbd/basic/qa/suites/rbd/basic)
```

Reply:
> @kshtsk I dont have much to say here except that it can have a more helpful name for folks who can print help and know better , 'calc' seems like generic name. 
I'm not against other convenient name, but would like to keep the logic.
> Also when I tried this command with default and -L it just prints the number of tests as shown below, so I strongly feel it should set those as defaults, Is there something I am missing here on usage?
Yes, you missed this:
```
-L, --list                  Print file list inovolved for building suite,
                               used only with -C and --desc arguments
```
So, if -P not provided, -L does not make sense, however, we can change it to 'if -L is used, -P is treated as true.'

Reply:
@varadakari I've addressed not working '-L' if '-P'.

> how about using a better name instead of --calc (like --analyze) ?

do you think it is really good name? :-) maybe `--analyse`  is better, or support both `--analyze` and `--analyse` as interchangeable arguments?